# Part 1. Equation of a Slime

How many late days are you using for this assignment?
0

In [8]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures


## 1. Loading the dataset

In [9]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
SDL_dataset = pd.read_csv('science_data_large.csv')
display(SDL_dataset.head(15))
display(SDL_dataset.info())

,Temperature °C,Mols KCL,Size nm^3
0,469,647,6.244743e+05
1,403,694,5.779610e+05
2,302,975,6.196847e+05
3,779,916,1.460449e+06
4,901,18,4.325726e+04
5,545,637,7.124634e+05
6,660,519,7.006960e+05
7,143,869,2.718260e+05
8,89,461,8.919803e+04
9,294,776,4.770210e+05


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


None

## 2. Splitting the dataset

In [10]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = SDL_dataset[['Temperature °C', 'Mols KCL']]
y = SDL_dataset['Size nm^3']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, train_size=0.9, random_state=42)

## 3. Perform a Linear Regression

In [11]:
# Use sklearn to train a model on the training set
reg = LinearRegression()
reg.fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
sample_point = pd.DataFrame([[469, 647]], columns=['Temperature °C', 'Mols KCL'])
prediction = reg.predict(sample_point)
print(f"Prediction for sample point: {prediction[0]:.2f} nm^3")
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
train_score = reg.score(X_train, y_train)
test_score = reg.score(X_test, y_test)
print(f"Training Score (R^2): {train_score:.4f}")
print(f"Testing Score (R^2): {test_score:.4f}")

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = reg.coef_
intercept = reg.intercept_
print(f"\nCoefficients: {coefficients}")
print(f"Intercept: {intercept}")


Prediction for sample point: 664984.90 nm^3
Training Score (R^2): 0.8610
Testing Score (R^2): 0.8552

Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340775


Write the linear equation of a slime: (example equation: $E = mc^2$)

equation of h(x) (rounded to 2 decimal points):
$$Size(nm^{3})= 866.15 \times Temperature(\degree C)+ 1032.70 \times Mols(KCL)−409391.48$$

Report on score and explain meaning:
- The training score and testing score are respectively 0.8610 and 0.8552 using R^2 formula as this is the default option in sklearn library. 
- The default score function in this context is actually the coefficient of determination, or we can also call it R-squared, will tell us how good the model perform based on the proportion of variation in the outcomes(labels) based on the features(inputs). To be specific, a score of around 0.86 in training means that about 86% of the variance in the size of slimes can be explained by temperature and mols of KCL. In other words, our model captures 86% of the relationship between these variables, with the remaining 14% being unexplained variation or noise in the data. For testing, the score of 0.8552 indicates that our model maintains similar predictive ability on unseen data, which means it generalizes well and isn't overfitting. The close alignment between training and testing scores (only about 0.6% difference) further supports the model's reliability.

## 4. Use Cross Validation

In [23]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
from sklearn.model_selection import cross_val_score, KFold

# Set up KFold cross-validation
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Perform cross-validation
cv_scores = cross_val_score(reg, X, y, cv=kf,scoring="r2")

# Display the cross-validation scores
print(f"Cross-validation scores (R^2): {cv_scores}")
print(f"Mean CV score: {cv_scores.mean():.4f}")
print(f"Standard deviation of CV scores: {cv_scores.std():.4f}")


Cross-validation scores (R^2): [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Mean CV score: 0.8597
Standard deviation of CV scores: 0.0184


## Cross-Validation Results Analysis

The cross-validation results strongly show that our model is relatively reliable and stable. We can observe that our model achieved a decent performance consistently across all 5 folds with R2 scores ranging from 0.8274 to 0.8817, with the mean of 5 folds around 0.8597. In other words, on average, our model explains about 86% of the variance in slime size based on temperature and KCL concentration. Our model also has a low standard deviation of 0.0201 which means there is only a small variation between fold scores. This is equivalent to the fact that our model performs consistently regardless of how the data is split in our case. When comparing with our initial train/test split, the mean CV score (0.8597) aligns closely with both our training (0.8610) and testing (0.8552) scores. This consistency across different validation methods confirms that our model can capture relationships in the data as a whole instead of getting over-fitting to specific data subsets. This cross-validation approach gives a more holistic and trustful evaluation than a single train/test split by utilizing the entire dataset while maintaining separation between training and validation data in each fold. The only problem is the potential of wasting computational resources and it may take significantly longer time based on the size of dataset.

## 5. Using Polynomial Regression

In [ ]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)
cv_scores = cross_val_score(reg, X_poly, y, cv=kf, scoring='r2')

print()
print(f"Cross-validation R2 scores for each fold: {cv_scores}")
print(f"Average R2 across all folds: {cv_scores.mean()}")

reg.fit(X_poly, y)
print(f"Intercept and coefficients of the model after: {reg.intercept_},{reg.coef_}")


# Print the intercept and coefs of the model
print(poly.get_feature_names_out(input_features=['Temperature °C', 'Mols KCL']))




Cross-validation MSE scores for each fold: [1. 1. 1. 1. 1.]
Average MSE across all folds: 1.0
Intercept and coefficients of the model: 1.657253596931696e-05,[ 0.00000000e+00  1.20000000e+01 -1.23112047e-07 -1.05657705e-11
  2.00000000e+00  2.85714287e-02]
Intercept and coefficients of the model after: 1.657253596931696e-05,[ 0.00000000e+00  1.20000000e+01 -1.23112047e-07 -1.05657705e-11
  2.00000000e+00  2.85714287e-02]
['1' 'Temperature °C' 'Mols KCL' 'Temperature °C^2'
 'Temperature °C Mols KCL' 'Mols KCL^2']


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

Polynomial equation of h(x) (degree 2):
$$Size(nm^3) = \beta_0 + \beta_1 \times Temperature(\degree C) + \beta_2 \times Mols(KCL) + \beta_3 \times Temperature(\degree C)^2 + \beta_4 \times Temperature(\degree C) \times Mols(KCL) + \beta_5 \times Mols(KCL)^2$$

This is equivalent to the following function:

$$Size(nm^3) = (1.657253596931696e-05) + (1.20000000e+01) \times Temperature(\degree C) + (-1.23112047e-07) \times Mols(KCL) + (-1.05657705e-11) \times Temperature(\degree C)^2 + ( 2.00000000e+00) \times Temperature(\degree C) \times Mols(KCL) + (2.85714287e-02) \times Mols(KCL)^2$$


Report on Score and Interpretation:

The R2 score of the model is exactly 1.0 for each fold which means the polynomial regression model is achieving a perfect fit - it explains 100% of the variance in the data across all cross-validation folds. This is a significant improvement from our previous linear model, which has score of about 0.86. The perfect score suggests that the relationship between our features (Temperature and Mols KCL) and the slime size follows a second-degree polynomial pattern very precisely. However, we also have to be careful as the score of 1.0 across all folds can sometimes indicate overfitting, though the consistency across folds suggests this might be a fit to the pattern of the data. The polynomial transformation allows the model to capture non-linear relationships between variables better, including interaction effects between Temperature and Mols KCL and quadratic terms of those features. The coefficients in our polynomial equation show how each term contributes to predicting slime size, with significant importance placed on the Temperature and the interaction term between Temperature and Mols KCL. The model suggests that slime size is influenced not only by individual factors but also by how these factors interact with each other, which wasn't captured in the linear model.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [57]:
import pandas as pd
from sklearn.model_selection import cross_val_score, KFold
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

# Load the dataset

ckd_data = pd.read_csv('ckd_feature_subset.csv')
print("Dataset shape:", ckd_data.shape)
display(ckd_data.head())

# Check for class distribution
print("Labels distribution:")
print(ckd_data['Target_ckd'].value_counts())

X = ckd_data.drop('Target_ckd', axis=1)
y = ckd_data['Target_ckd']

# Scale the features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Set up cross-validation
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Train and evaluate multiple classification models
models = {
    'Logistic Regression': LogisticRegression(random_state=42),
    'SVM': SVC(random_state=42),
    'k-NN': KNeighborsClassifier(),
    'Neural Network': MLPClassifier(hidden_layer_sizes=(10, 5), random_state=42, max_iter=1000)
}

# store results
results = {}

# Evaluate each model with cross-validation
for name, model in models.items():
    cv_scores = cross_val_score(model, X_scaled, y, cv=kf, scoring='accuracy')
    results[name] = {
        'scores': cv_scores,
        'mean': cv_scores.mean(),
        'std': cv_scores.std()
    }
    print(f"{name} - Mean accuracy: {cv_scores.mean():.4f}, Std: {cv_scores.std():.4f}")

# Create a results table
results_df = pd.DataFrame({
    'Model': list(results.keys()),
    'Mean Accuracy': [results[model]['mean'] for model in models],
    'Std Dev': [results[model]['std'] for model in models]
})

# Sort by performance
display(results_df.sort_values('Mean Accuracy', ascending=False))


Dataset shape: (153, 7)


,age,bp,wbcc,appet_poor,appet_good,rbcc,Target_ckd
0,0.688312,0.333333,0.000000,1,0,0.000000,1
1,0.545455,0.333333,0.128319,1,0,0.305085,1
2,0.714286,0.500000,0.238938,1,0,0.186441,1
3,0.688312,0.333333,0.283186,0,1,0.338983,1
4,0.441558,0.333333,0.221239,1,0,0.220339,1


Labels distribution:
Target_ckd
0    111
1     42
Name: count, dtype: int64
Logistic Regression - Mean accuracy: 0.9480, Std: 0.0436
SVM - Mean accuracy: 0.9671, Std: 0.0293
k-NN - Mean accuracy: 0.9215, Std: 0.0604
Neural Network - Mean accuracy: 0.9675, Std: 0.0353


,Model,Mean Accuracy,Std Dev
3,Neural Network,0.967527,0.035340
1,SVM,0.967097,0.029342
0,Logistic Regression,0.947957,0.043569
2,k-NN,0.921505,0.060430


## Results and Conclusion for Classification Experiments

### Dataset Analysis
The CKD dataset contains 153 samples with 6 features (age, blood pressure, white blood cell count, appetite status, and red blood cell count) and a binary target variable indicating chronic kidney disease. The class distribution is imbalanced with 111 negative cases (72.5%) and 42 positive cases (27.5%). All features were standardized before model training.

### Model Performance Analysis
Four classification models were evaluated using 5-fold cross-validation to ensure robust performance assessment:

| Model | Mean Accuracy | Standard Deviation |
|-------|---------------|-------|
| Neural Network | 96.75% | 3.53% |
| SVM | 96.71% | 2.93% |
| Logistic Regression | 94.80% | 4.36% |
| k-NN | 92.15% | 6.04% |

Both the Neural Network (MLP with two hidden layers) and SVM have excellent performance with mean accuracies above 96.7%. The SVM model shows slightly more consistent performance across 5 folds with the lowest standard deviation (2.93%) while the NN has a very small better in mean score. Logistic Regression performed well with nearly 95% accuracy, while k-NN had the lowest performance and highest variance (not very stable) among the 4 models.

### Results Analysis
1. **SVM** has a strong performance thanks to its effectiveness in finding optimal decision boundaries in higher-dimensional spaces, especially for smaller datasets like this one.

2. **Neural Networks** is as good as SVM performance, showing that they perform well even with limited data. Usually, it is a problem to appropriately  balance between model complexity and generalization(which potentially lead to under-fitting and over-fitting based on models). So, this simple architecture has proved to work great.

3. **Logistic Regression's** a relative solid performance indicates with the decision boundaries classify the classes well enough.

4. **k-NN's** relatively lower performance suggests that simple distance-based classification is less effective for this dataset, possibly due to overlapping in class distributions or inappropriate dimensionality.

5. **Low standard deviations** across most models (except k-NN which is somewhat higher than the others) indicate that the performance is consistent across different data splits, suggesting that the models can capture patterns of the data.
### Overall
Although we are facing classes imbalance issue, all models still achieved high accuracy, indicating that the features provide strong predictive signals for kidney disease classification. The high performance across different model types suggests that the dataset contains well-separable classes once the features are properly scaled.

In my opinion, the SVM is thee best model based on our experiment even though the NN show a slightly improvement in the accuracy but the difference in the average mean is too trivial (only 0.04%) but the difference in standard deviation (around 0.6%) outweighs its advantage in accuracy. Hence I believe SVM is the best model for being one of the models has highest accuracy score and lowest deviation.

In [54]:
# Experiments with Neural Network.
nn_configs = [
    {
        'name': 'Single hidden layer (10 neurons)',
        'model': MLPClassifier(hidden_layer_sizes=(10,), random_state=42, max_iter=1000)
    },
    {
        'name': 'Two hidden layers (10, 5)',
        'model': MLPClassifier(hidden_layer_sizes=(10, 5), random_state=42, max_iter=1000)
    },
    {
        'name': 'Large hidden layer (50)',
        'model': MLPClassifier(hidden_layer_sizes=(50,), random_state=42, max_iter=1000)
    },
    {
        'name': 'Deep network (20, 10, 5)',
        'model': MLPClassifier(hidden_layer_sizes=(20, 10, 5), random_state=42, max_iter=1000)
    },
    {
        'name': 'With regularization (alpha=0.01)',
        'model': MLPClassifier(hidden_layer_sizes=(10, 5), alpha=0.01, random_state=42, max_iter=1000)
    }
]

# Dictionary to store neural network results
nn_results = {}

# Evaluate each neural network configuration
for config in nn_configs:
    name = config['name']
    model = config['model']
    
    cv_scores = cross_val_score(model, X_scaled, y, cv=kf, scoring='accuracy')
    
    nn_results[name] = {
        'scores': cv_scores,
        'mean': cv_scores.mean(),
        'std': cv_scores.std()
    }
    
    print(f"{name} - Mean accuracy: {cv_scores.mean():.4f}, Std: {cv_scores.std():.4f}")

# Create a DataFrame for neural network results
nn_results_df = pd.DataFrame({
    'Configuration': list(nn_results.keys()),
    'Mean Accuracy': [nn_results[config]['mean'] for config in nn_results],
    'Std Dev': [nn_results[config]['std'] for config in nn_results]
})

display(nn_results_df.sort_values('Mean Accuracy', ascending=False))

c:\Users\tuant\AppData\Local\Programs\Python\Python38\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\tuant\AppData\Local\Programs\Python\Python38\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\tuant\AppData\Local\Programs\Python\Python38\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\tuant\AppData\Local\Programs\Python\Python38\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the opti

Single hidden layer (10 neurons) - Mean accuracy: 0.9544, Std: 0.0436
Two hidden layers (10, 5) - Mean accuracy: 0.9675, Std: 0.0353
Large hidden layer (50) - Mean accuracy: 0.9675, Std: 0.0353
Deep network (20, 10, 5) - Mean accuracy: 0.9740, Std: 0.0377
With regularization (alpha=0.01) - Mean accuracy: 0.9675, Std: 0.0353


,Configuration,Mean Accuracy,Std Dev
3,"Deep network (20, 10, 5)",0.973978,0.037658
1,"Two hidden layers (10, 5)",0.967527,0.035340
2,Large hidden layer (50),0.967527,0.035340
4,With regularization (alpha=0.01),0.967527,0.035340
0,Single hidden layer (10 neurons),0.954409,0.043633


## Results and Conclusion for Neural Network Experiments

### Neural Network Architecture Comparison

We experimented with five different neural network configurations to determine the optimal architecture for CKD prediction:

| Configuration | Mean Accuracy | Standard Deviation |
|---------------|---------------|-------------------|
| Deep network (20, 10, 5) | 97.40% | 3.77% |
| Two hidden layers (10, 5) | 96.75% | 3.53% |
| Large hidden layer (50) | 96.75% | 3.53% |
| With regularization (alpha=0.01) | 96.75% | 3.53% |
| Single hidden layer (10) | 95.44% | 4.36% |

### Important Findings

1. **Architecture**: The deep network with three hidden layers (20, 10, 5) achieved the highest accuracy (97.40%), suggesting that a deeper architecture can better capture the complex patterns in the CKD dataset.

2. **Width and Depth**: The single large hidden layer with 50 neurons performed identically to the two-layer network (10, 5), indicating that simply adding more neurons to a single layer doesn't necessarily improve performance over a thoughtfully structured multi-layer network.

3. **Regularization Impact**: Adding regularization (alpha=0.01) to the two-layer network didn't change performance, which means that overfitting is not a significant issue with our model and the dataset.

4. **Stability**: Generally, the standard deviation is stable among all configs. Though, The deeper architectures showed slightly higher deviation in performance across folds, which is expected as more complex models can be more sensitive to changes in data.

5. **Simplicity and Performance**: While the single hidden layer model performed worse than the others, its 95.44% accuracy is still quite good, demonstrating that even a simple neural network can effectively capture most of the pattern in this dataset.

### Overall

The results suggest that for this CKD dataset, Deep Network will hold best performance in accuracy but slightly higher deviation compared to other configs. However, the performance differences between the different architectures are relatively small (less than 2% between best and worst so they all nearly perform the same), suggesting that the features in this dataset are very effective to predict the target regardless of the exact neural network configuration.

In real life applications, I believe two-layer network (10, 5) is enough to offer an excellent balance between performance (96.75% accuracy) and computational simplicity compared to deeper architectures, so this one is the best option in my opinion.
